tao_compiler/mlir/disc/transforms/fusion_utils.cc

```diff
@@ -2521,5 +2822,1033 @@ bool StitchCPUAnalysis::doCodeGeneration(OpBuilder& b, lmhlo::FusionOp fusion) {
+}
+
+bool StitchGpuFusionStrategy::tryFuse(ShapeAnalysis& shapeAnalysis,
+                                      FusionPattern& lhs, FusionPattern& rhs,
+                                      FusionPattern& target) {
+  if (!FusionStrategy::tryFuse(shapeAnalysis, lhs, rhs, target)) {
+    return false;
+  }
+
+  return initFusionPattern(shapeAnalysis, target);
+}
+
+bool StitchGpuFusionStrategy::tryFuseInplace(ShapeAnalysis& shapeAnalysis,
+                                             FusionPattern& lhs,
+                                             FusionPattern& rhs) {
+  // both lhs & rhs should be fusible
+  if (!isFusible(lhs) || !isFusible(rhs)) {
+    return false;
+  }
+  FusionPattern result = lhs.mergeWithoutInit(rhs);
+  if (!tryFuse(shapeAnalysis, lhs, rhs, result)) {
+    return false;
+  }
+  lhs = result;
+  return true;
+}
+
+bool StitchGpuFusionStrategy::findFusionPatternTypeAndSubroot(
+    ShapeAnalysis& shapeAnalysis, FusionPattern& fusion_pattern) {
+  FusionType fusion_type = FusionType::kNone;
+  Operation* dominant_op = nullptr;
+  auto results = fusion_pattern.getResults();
+  if (results.empty()) {
+    return false;
+  }
+
+  SmallVector<Operation*, 4> row_reductions;
+  SmallVector<Operation*, 4> col_reductions;
+  if (!findValidReductionOps(fusion_pattern, row_reductions, col_reductions)) {
+    LLVM_DEBUG(llvm::dbgs() << "Check reduction ops failed.");
+    return false;
+  }
+
+  // Check whether there is 'middle' row reduction. If there is, it is kStitch.
+  auto op_list = fusion_pattern.getOpList();
+  DenseSet<Operation*> op_set(op_list.begin(), op_list.end());
+  for (Operation* op : row_reductions) {
+    if (fusion_type != FusionType::kNone) {
+      break;
+    }
+    int num_input_operand = op->getNumOperands() - getNumResultOperands(op);
+    for (Value v : op->getOperands().drop_front(num_input_operand)) {
+      if (fusion_type != FusionType::kNone) {
+        break;
+      }
+      for (Operation* user : getValueUsers(v)) {
+        if (user == op) continue;
+        if (op_set.find(user) != op_set.end()) {
+          fusion_type = FusionType::kStitch;
+          dominant_op = row_reductions.back();
+          break;
+        }
+      }
+    }
+  }
+
+  // Check basic fusion types.
+  // Check compatibility for kRowReduction, kColReduction and kLoop:
+  // - for kRowReduction, all row-reduces should have the same shape with
+  // dominant, and other results should have the same number of elements.
+  // Otherwise we set it as kStitch fusion.
+  // - for kColReduction, all col-reduces should have the same shape with
+  // dominant, and other results should have the same number of elements.
+  // - for kLoop, all results should have the same number of elements.
+  // Note that kRowReduce that do not meet compatibility constraint are already
+  // regarded as kStitch already.
+  if (fusion_type != FusionType::kStitch) {
+    if (!row_reductions.empty()) {
+      fusion_type = FusionType::kRowReduction;
```

Comment:
Since a 'base' strategy is already run ahead of 'stitch' strategy. Why do we still need to handle non-stitch fusion types here? There shouldn't be any more fusion space except for stitch fusion?

Reply:
Because run base strategy before stitch strategy may make the optimization space not able to reach. The current implement is to save compiling time. This design supports to run stitch strategy directly without any other pre-processing.